web/client/utils/KeycloakUtils.js

```diff
@@ -0,0 +1,238 @@
+            login() {
+                // check if is not logged in or if is logging in via openid
+                if (!isLoggedIn(store.getState()) && !isLoggingIn()) {
+                    commandSubject.next("login");
+                // token for keycloak is not initialized. Reinit the client to set the token.
+                } else if (!keycloak.authenticated && isLoggedInWithSSOProvider() ) {
+                    commandSubject.next("syncToken"); // TODO: this should be triggered only once, or find a different way.
+                }
+                return Promise.resolve();
+            },
+            logout() {
+                // logout only if you are logged in
+                if (isLoggedInWithSSOProvider()) {
+                    clearKeycloakSession(keycloak);
+                    commandSubject.next("logout");
+
+                }
+                return Promise.resolve();
+            },
+            redirectUri(options) {
+                if (options && options.redirectUri) {
+                    return options.redirectUri;
+                } else if (keycloak.redirectUri) {
+                    return keycloak.redirectUri;
+                }
+                return location.href;
+            }
+        }
+    }).then(() => {
+        if (keycloak.authenticated) {
+            if (isLoggedInWithSSOProvider() && keycloak?.tokenParsed?.exp && keycloak?.refreshTokenParsed?.exp) {
+                commandSubject.next("scheduleRefresh");
+            }
+        // not logged in keycloak or in MapStore session, schedule login monitoring.
+        } else if (!keycloak.authenticated && !isLoggedIn(store.getState())) {
+            commandSubject.next("noSessionFound");
+        }
+    })
+        .catch((e) => {
+            console.error("Error initializing keycloak SSO", e);
+            clearKeycloakSession(keycloak);
+            commandSubject.next("logout");
+        });
+
+
+    keycloak.onAuthLogout = () => {
+        clearKeycloakSession(keycloak);
+        commandSubject.next("logout");
+    };
+    return commandSubject.asObservable();
+};
+
+/**
+ * Monitors the SSO login status of the user using the Keycloak library.
+ * @param {object} ssoProvider the sso provider
+ * @param {object} store the redux store
+ * @returns {Observable} the observable that emits the login/logout redux actions
+ */
+export const monitorKeycloak = (ssoProvider) => (action$, store) => {
+    const initSubject = new Rx.Subject(); // commands emitted here are "retry" and "syncToken"
+    return Rx.Observable.of(ssoProvider)
+        // re-init on login success to set up the token
+        .merge(
+            action$.ofType(LOGIN_SUCCESS)
+                .filter(({authProvider}) => authProvider === ssoProvider.provider)
+        )
+        .merge(
+            action$.ofType(REFRESH_SUCCESS)
+            // .filter(({authProvider}) => authProvider === ssoProvider.provider) // refresh auth provider is geostore. so can not filter.
```

Comment:
Why this is commented? Is it needed action$.ofType(REFRESH_SUCCESS) ?

Reply:
Current refresh is applied by `geostore`. so the provider will be `geostore`. 
so I can remove the commented code and put a note. 
```suggestion
```